File: internal/sources/pi.go
package sources

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"github.com/HelloiOS2014/AgentRecovery/internal/core"
)

var piUUID = regexp.MustCompile(`[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}`)

func defaultPiSessions() string {
	if v := os.Getenv("PI_AGENT_DIR"); v != "" {
		return filepath.Join(v, "sessions")
	}
	home, _ := os.UserHomeDir()
	return filepath.Join(home, ".pi", "agent", "sessions")
}

func piUUIDFromName(name string) string {
	stem := name
	if strings.HasSuffix(name, ".jsonl") {
		stem = name[:len(name)-6]
	}
	return strings.ToLower(piUUID.FindString(stem))
}

type Pi struct {
	SessionsDir string
}

func NewPi(dir string) *Pi {
	if dir == "" {
		dir = defaultPiSessions()
	}
	return &Pi{SessionsDir: dir}
}

func (p *Pi) Name() string                  { return "pi" }
func (p *Pi) LoadTitles() map[string]string { return map[string]string{} }

func (p *Pi) sessionFiles() (map[string]string, error) {
	found := map[string]struct {
		path  string
		mtime int64
	}{}
	st, err := os.Stat(p.SessionsDir)
	if err != nil {
		if os.IsNotExist(err) {
			return map[string]string{}, nil
		}
		return nil, err
	}
	if !st.IsDir() {
		return map[string]string{}, nil
	}
	err = filepath.WalkDir(p.SessionsDir, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if d.IsDir() || !strings.HasSuffix(d.Name(), ".jsonl") {
			return nil
		}
		sid := piUUIDFromName(d.Name())
		if sid == "" {
			return nil
		}
		info, err := d.Info()
		if err != nil {
			return nil
		}
		mt := info.ModTime().UnixNano()
		if old, ok := found[sid]; !ok || mt > old.mtime {
			found[sid] = struct {
				path  string
				mtime int64
			}{path, mt}
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	out := map[string]string{}
	for sid, h := range found {
		out[sid] = h.path
	}
	return out, nil
}

func contentText(content any) (string, int) {
	if s, ok := content.(string); ok {
		return s, 0
	}
	nImg := 0
	var parts []string
	for _, b := range asList(content) {
		m := asMap(b)
		switch asStr(m["type"]) {
		case "text":
			parts = append(parts, asStr(m["text"]))
		case "image":
			nImg++
		}
	}
	return strings.Join(parts, "\n"), nImg
}

func (p *Pi) ListSessions(limit int) ([]core.SessionMeta, error) {
	files, err := p.sessionFiles()
	if err != nil {
		return nil, err
	}
	type pair struct {
		id, path string
		mtime    int64
	}
	var order []pair
	for sid, path := range files {
		info, err := os.Stat(path)
		if err != nil {
			continue
		}
		order = append(order, pair{sid, path, info.ModTime().UnixNano()})
	}
	sort.Slice(order, func(i, j int) bool { return order[i].mtime > order[j].mtime })
	if limit > 0 && len(order) > limit {
		order = order[:limit]
	}
	var metas []core.SessionMeta
	for _, x := range order {
		m := p.scanMeta(x.path, x.id)
		info, _ := os.Stat(x.path)
		m.UpdatedAt = isoSec(info.ModTime())
		metas = append(metas, m)
	}
	return metas, nil
}

func (p *Pi) scanMeta(path, sid string) core.SessionMeta {
	meta := core.SessionMeta{ID: sid}
	models := map[string]bool{}
	rec, _, err := readJSONL(path)
	if err != nil {
		return meta
	}
	for _, r := range rec {
		switch asStr(r["type"]) {
		case "session":
			if asStr(r["cwd"]) != "" {
				meta.Cwd = asStr(r["cwd"])
			}
			if asStr(r["timestamp"]) != "" {
				meta.StartedAt = asStr(r["timestamp"])
			}
			if asStr(r["id"]) != "" {
				meta.ID = asStr(r["id"])
			}
		case "session_info":
			if n := asStr(r["name"]); n != "" {
				meta.Title = n
			}
		case "model_change":
			if m := asStr(r["modelId"]); m != "" {
				models[m] = true
			}
		case "message":
			msg := asMap(r["message"])
			if m := asStr(msg["model"]); m != "" {
				models[m] = true
			}
			if meta.Title == "" && asStr(msg["role"]) == "user" {
				text, _ := contentText(msg["content"])
				text = strings.TrimSpace(text)
				if text != "" {
					meta.Title = firstLine(text, 80)
				}
			}
		}
	}
	if len(models) > 0 {
		meta.Model = strings.Join(sortedKeys(models), ", ")
	}
	return meta
}

func firstLine(s string, n int) string {
	if i := strings.IndexByte(s, '\n'); i >= 0 {
		s = s[:i]
	}
	r := []rune(s)
	if len(r) > n {
		return string(r[:n])
	}
	return s
}

func (p *Pi) ReadSession(id string) (*core.Session, error) {
	want := strings.ToLower(id)
	files, err := p.sessionFiles()
	if err != nil {
		return nil, err
	}
	path := files[want]
	if path == "" {
		for sid, fp := range files {
			if strings.HasPrefix(sid, want) || strings.HasPrefix(want, sid) {
				path = fp
				break
			}
		}
	}
	if path == "" {
		return nil, fmt.Errorf("%w: 未找到 Pi 会话 %s（已扫描 %s）", ErrNotFound, id, p.SessionsDir)
	}
	return p.parse(path, id)
}

func (p *Pi) leafPath(records []map[string]any) []map[string]any {
	var tree []map[string]any
	for _, r := range records {
		if asStr(r["type"]) == "session" || asStr(r["id"]) == "" {
			continue
		}
		tree = append(tree, r)
	}
	if len(tree) == 0 {
		return nil
	}
	allNil := true
	for _, r := range tree {
		if asStr(r["parentId"]) != "" {
			allNil = false
			break
		}
	}
	if allNil {
		return tree
	}
	byID := map[string]map[string]any{}
	children := map[string]bool{}
	for _, r := range tree {
		id := asStr(r["id"])
		byID[id] = r
		if pid := asStr(r["parentId"]); pid != "" {
			children[pid] = true
		}
	}
	var leaves []map[string]any
	for _, r := range tree {
		if !children[asStr(r["id"])] {
			leaves = append(leaves, r)
		}
	}
	leaf := tree[len(tree)-1]
	if len(leaves) > 0 {
		leaf = leaves[0]
		for _, r := range leaves[1:] {
			if asStr(r["timestamp"]) > asStr(leaf["timestamp"]) {
				leaf = r
			}
		}
	}
	var path []map[string]any
	seen := map[string]bool{}
	cur := asStr(leaf["id"])
	for cur != "" && byID[cur] != nil && !seen[cur] {
		path = append(path, byID[cur])
		seen[cur] = true
		cur = asStr(byID[cur]["parentId"])
	}
	for i, j := 0, len(path)-1; i < j; i, j = i+1, j-1 {
		path[i], path[j] = path[j], path[i]
	}
	return path
}

func afterCompaction(path []map[string]any) (rest []map[string]any, tail []any, compacted bool, summary string) {
	last := -1
	for i, r := range path {
		if asStr(r["type"]) == "compaction" {
			last = i
		}
	}
	if last < 0 {
		return path, nil, false, ""
	}
	rec := path[last]
	summary = asStr(rec["summary"])
	if t, ok := rec["retainedTail"].([]any); ok {
		tail = t
	}
	for _, r := range path[last+1:] {
		if asStr(r["type"]) != "compaction" {
			rest = append(rest, r)
		}
	}
	return rest, tail, true, summary
}

func (p *Pi) parse(path, id string) (*core.Session, error) {
	meta := core.SessionMeta{ID: id, Source: "pi"}
	var events []core.Event
	calls := map[string]int{}
	var warnings []string
	models := map[string]bool{}
	unknown := map[string]int{}
	skippedImg := 0

	records, bad, err := readJSONL(path)
	if err != nil {
		return nil, err
	}
	for _, r := range records {
		if asStr(r["type"]) == "session" {
			meta.Cwd = asStr(r["cwd"])
			meta.StartedAt = asStr(r["timestamp"])
			if asStr(r["id"]) != "" {
				meta.ID = asStr(r["id"])
			}
			break
		}
	}
	pathRecs := p.leafPath(records)
	rest, retained, compacted, summary := afterCompaction(pathRecs)
	if compacted {
		warnings = append(warnings, "会话压缩过：已重置到压缩边界，仅保留摘要之后的上下文")
		if summary != "" {
			s := summary
			if len([]rune(s)) > 200 {
				s = string([]rune(s)[:200]) + "…"
			}
			warnings = append(warnings, "压缩摘要："+s)
		}
	}
	for _, msg := range retained {
		if m := asMap(msg); m != nil {
			skippedImg += handlePiMessage(m, &events, calls, &warnings, models)
		}
	}
	for _, r := range rest {
		switch asStr(r["type"]) {
		case "message":
			skippedImg += handlePiMessage(asMap(r["message"]), &events, calls, &warnings, models)
		case "session_info":
			if n := asStr(r["name"]); n != "" {
				meta.Title = n
			}
		case "model_change":
			if m := asStr(r["modelId"]); m != "" {
				models[m] = true
			}
		case "thinking_level_change", "custom", "custom_message", "label", "branch_summary", "compaction":
		default:
			t := asStr(r["type"])
			if t != "" {
				unknown[t]++
			}
		}
	}
	if meta.Title == "" {
		for _, e := range events {
			if e.Kind == "user_msg" && strings.TrimSpace(e.Text) != "" {
				meta.Title = firstLine(strings.TrimSpace(e.Text), 80)
				break
			}
		}
	}
	if len(models) > 0 {
		meta.Model = strings.Join(sortedKeys(models), ", ")
	}
	if bad > 0 {
		warnings = append(warnings, fmt.Sprintf("解析中跳过 %d 个坏行（并发写入或中断所致）", bad))
	}
	if skippedImg > 0 {
		warnings = append(warnings, fmt.Sprintf("跳过 %d 张图片（不恢复附件）", skippedImg))
	}
	keys := sortedKeys(func() map[string]bool {
		m := map[string]bool{}
		for t := range unknown {
			m[t] = true
		}
		return m
	}())
	for _, t := range keys {
		warnings = append(warnings, fmt.Sprintf("跳过未知记录类型 %s（%d 条）", t, unknown[t]))
	}
	return &core.Session{Meta: meta, Events: events, Compacted: compacted, Warnings: warnings}, nil
}

func handlePiMessage(msg map[string]any, events *[]core.Event, calls map[string]int, warnings *[]string, models map[string]bool) int {
	nImg := 0
	if m := asStr(msg["model"]); m != "" {
		models[m] = true
	}
	switch asStr(msg["role"]) {
	case "user":
		text, n := contentText(msg["content"])
		nImg += n
		if strings.TrimSpace(text) != "" {
			*events = append(*events, core.Event{Kind: "user_msg", Role: "user", Text: strings.TrimSpace(text)})
		}
	case "assistant":
		content := msg["content"]
		if list := asList(content); list != nil {
			for _, b := range list {
				m := asMap(b)
				switch asStr(m["type"]) {
				case "thinking":
					th := strings.TrimSpace(asStr(m["thinking"]))
					if th != "" {
						*events = append(*events, core.Event{Kind: "reasoning", Role: "assistant", Text: th})
					}
				case "text":
					t := strings.TrimSpace(asStr(m["text"]))
					if t != "" {
						*events = append(*events, core.Event{Kind: "assistant_msg", Role: "assistant", Text: t})
					}
				case "toolCall":
					*events = append(*events, core.Event{Kind: "tool_call", Text: orQ(asStr(m["name"])), ToolArgs: jsonDump(m["arguments"])})
					if id := asStr(m["id"]); id != "" {
						calls[id] = len(*events) - 1
					}
				case "image":
					nImg++
				}
			}
		} else if s, ok := content.(string); ok && strings.TrimSpace(s) != "" {
			*events = append(*events, core.Event{Kind: "assistant_msg", Role: "assistant", Text: strings.TrimSpace(s)})
		}
	case "toolResult":
		text, n := contentText(msg["content"])
		nImg += n
		tid := asStr(msg["toolCallId"])
		if idx, ok := calls[tid]; ok {
			(*events)[idx].ToolOutput = text
		} else {
			*events = append(*events, core.Event{Kind: "tool_output", Text: text})
			*warnings = append(*warnings, "存在无法配对 toolCallId 的工具输出（已顺序追加）")
		}
	case "bashExecution":
		*events = append(*events, core.Event{
			Kind: "tool_call", Text: "bash",
			ToolArgs: asStr(msg["command"]), ToolOutput: asStr(msg["output"]),
		})
	}
	return nImg
}

File: scripts/sources/pi.py
"""PiSource: discover and parse Pi session files.

Facts verified on real data (2026-08-31):
- Sessions: ~/.pi/agent/sessions/--<cwd-with-/-as--->--/<ISO-ts>_<uuid>.jsonl
- Header: {type: session, version: 3, id, timestamp, cwd} — not part of the tree
- Tree via id/parentId; recover the current leaf path (latest timestamp among
  entries with no children). Linear files (no parentId) keep record order.
- Assistant content blocks: text / thinking / toolCall (id, name, arguments)
- Tool results are separate messages, role=toolResult, paired by toolCallId
- Compaction entries reset context like Claude compact_boundary
"""
import json
import os
import re
from datetime import datetime
from typing import Dict, List, Optional, Set, Tuple

from . import Event, Session, SessionMeta, Source

UUID_RE = re.compile(r"[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}", re.I)


def default_sessions_dir() -> str:
    agent = os.environ.get("PI_AGENT_DIR") or os.path.join(os.path.expanduser("~"), ".pi", "agent")
    return os.path.join(agent, "sessions")


def _uuid_from_filename(name: str) -> Optional[str]:
    stem = name[:-6] if name.endswith(".jsonl") else name
    m = UUID_RE.search(stem)
    return m.group(0).lower() if m else None


def _content_text(content) -> Tuple[str, int]:
    if isinstance(content, str):
        return content, 0
    if not isinstance(content, list):
        return "", 0
    parts, n_img = [], 0
    for b in content:
        if not isinstance(b, dict):
            continue
        if b.get("type") == "text" and isinstance(b.get("text"), str):
            parts.append(b["text"])
        elif b.get("type") == "image":
            n_img += 1
    return "\n".join(parts), n_img


def _iter_records(path: str) -> Tuple[List[dict], int]:
    records, bad = [], 0
    with open(path) as fh:
        for line in fh:
            line = line.strip()
            if not line:
                continue
            try:
                records.append(json.loads(line))
            except ValueError:
                bad += 1
    return records, bad


class PiSource(Source):
    name = "pi"

    def __init__(self, sessions_dir: Optional[str] = None) -> None:
        self.sessions_dir = sessions_dir or default_sessions_dir()

    def _session_files(self) -> Dict[str, str]:
        found: Dict[str, Tuple[str, float]] = {}
        if not os.path.isdir(self.sessions_dir):
            return {}
        for root, _, files in os.walk(self.sessions_dir):
            for f in files:
                if not f.endswith(".jsonl"):
                    continue
                sid = _uuid_from_filename(f)
                if not sid:
                    continue
                path = os.path.join(root, f)
                mtime = os.path.getmtime(path)
                if sid not in found or mtime > found[sid][1]:
                    found[sid] = (path, mtime)
        return {sid: pair[0] for sid, pair in found.items()}

    def _scan_meta(self, path: str, sid: str) -> SessionMeta:
        title = cwd = started_at = model = None
        records, _ = _iter_records(path)
        models: Set[str] = set()
        for r in records:
            t = r.get("type")
            if t == "session":
                cwd = r.get("cwd") or cwd
                started_at = r.get("timestamp") or started_at
                if r.get("id"):
                    sid = r["id"]
            elif t == "session_info" and r.get("name"):
                title = r["name"]
            elif t == "model_change" and r.get("modelId"):
                models.add(r["modelId"])
            elif t == "message":
                msg = r.get("message") or {}
                if msg.get("model"):
                    models.add(msg["model"])
                if title is None and msg.get("role") == "user":
                    text, _ = _content_text(msg.get("content"))
                    text = text.strip()
                    if text:
                        title = text.splitlines()[0][:80]
        if models:
            model = ", ".join(sorted(models))
        return SessionMeta(id=sid, title=title, cwd=cwd, started_at=started_at, model=model)

    def list_sessions(self, limit: int = 20) -> List[SessionMeta]:
        files = self._session_files()
        if not files:
            return []
        order = sorted(files.items(), key=lambda kv: os.path.getmtime(kv[1]), reverse=True)
        metas = []
        for sid, path in order[:limit]:
            meta = self._scan_meta(path, sid)
            meta.updated_at = datetime.fromtimestamp(os.path.getmtime(path)).isoformat(timespec="seconds")
            metas.append(meta)
        return metas

    def _find_file(self, session_id: str) -> Optional[str]:
        want = session_id.lower()
        files = self._session_files()
        if want in files:
            return files[want]
        # partial / exact header id mismatch vs filename
        for sid, path in files.items():
            if sid.startswith(want) or want.startswith(sid):
                return path
        return None

    def read_session(self, session_id: str) -> Session:
        path = self._find_file(session_id)
        if not path:
            raise LookupError(
                "未找到 Pi 会话 %s（已扫描 %s）。\n"
                "会话 ID 来自 `~/.pi/agent/sessions/` 文件名或 `/session` 显示的 ID。"
                % (session_id, self.sessions_dir))
        return self._parse(path, session_id)

    def _leaf_path(self, records: List[dict]) -> List[dict]:
        tree = [r for r in records if r.get("type") != "session" and r.get("id")]
        if not tree:
            return []
        if all(not r.get("parentId") for r in tree):
            return tree
        by_id = {r["id"]: r for r in tree}
        children: Set[str] = set()
        for r in tree:
            pid = r.get("parentId")
            if pid:
                children.add(pid)
        leaves = [r for r in tree if r["id"] not in children]
        if not leaves:
            leaf = tree[-1]
        else:
            leaf = max(leaves, key=lambda r: r.get("timestamp") or "")
        path, seen = [], set()
        cur = leaf.get("id")
        while cur and cur in by_id and cur not in seen:
            path.append(by_id[cur])
            seen.add(cur)
            cur = by_id[cur].get("parentId")
        path.reverse()
        return path

    def _after_compaction(self, path: List[dict]) -> Tuple[List[dict], Optional[list], bool, Optional[str]]:
        last = None
        for i, r in enumerate(path):
            if r.get("type") == "compaction":
                last = i
        if last is None:
            return path, None, False, None
        rec = path[last]
        summary = rec.get("summary")
        tail = rec.get("retainedTail") if isinstance(rec.get("retainedTail"), list) else None
        rest = [r for r in path[last + 1 :] if r.get("type") != "compaction"]
        return rest, tail, True, summary

    def _parse(self, path: str, session_id: str) -> Session:
        meta = SessionMeta(id=session_id, source="pi")
        events: List[Event] = []
        calls: Dict[str, int] = {}
        warnings: List[str] = []
        models: Set[str] = set()
        unknown: Dict[str, int] = {}
        skipped_images = 0

        records, bad = _iter_records(path)
        for r in records:
            if r.get("type") == "session":
                meta.cwd = r.get("cwd") or meta.cwd
                meta.started_at = r.get("timestamp") or meta.started_at
                if r.get("id"):
                    meta.id = r["id"]
                break

        path_recs = self._leaf_path(records)
        rest, retained, compacted, summary = self._after_compaction(path_recs)
        if compacted:
            warnings.append("会话压缩过：已重置到压缩边界，仅保留摘要之后的上下文")
            if summary:
                warnings.append("压缩摘要：%s" % (summary[:200] + ("…" if len(summary) > 200 else "")))
        if retained:
            for msg in retained:
                if isinstance(msg, dict):
                    skipped_images += self._handle_message(msg, events, calls, warnings, models)

        for r in rest:
            t = r.get("type")
            if t == "message":
                skipped_images += self._handle_message(r.get("message") or {}, events, calls, warnings, models)
            elif t == "session_info":
                if r.get("name"):
                    meta.title = r["name"]
            elif t == "model_change":
                if r.get("modelId"):
                    models.add(r["modelId"])
            elif t in ("thinking_level_change", "custom", "custom_message",
                       "label", "branch_summary", "compaction"):
                pass
            else:
                unknown[t or "?"] = unknown.get(t or "?", 0) + 1

        if meta.title is None:
            for e in events:
                if e.kind == "user_msg" and (e.text or "").strip():
                    meta.title = e.text.strip().splitlines()[0][:80]
                    break
        if models:
            meta.model = ", ".join(sorted(models))
        if bad:
            warnings.append("解析中跳过 %d 个坏行（并发写入或中断所致）" % bad)
        if skipped_images:
            warnings.append("跳过 %d 张图片（不恢复附件）" % skipped_images)
        for t, n in sorted(unknown.items()):
            warnings.append("跳过未知记录类型 %s（%d 条）" % (t, n))
        return Session(meta=meta, events=events, compacted=compacted, warnings=warnings)

    @staticmethod
    def _handle_message(msg, events, calls, warnings, models) -> int:
        n_img = 0
        role = msg.get("role")
        if msg.get("model"):
            models.add(msg["model"])
        if role == "user":
            text, n_img = _content_text(msg.get("content"))
            if text.strip():
                events.append(Event(kind="user_msg", role="user", text=text.strip()))
        elif role == "assistant":
            content = msg.get("content")
            if isinstance(content, list):
                for b in content:
                    if not isinstance(b, dict):
                        continue
                    bt = b.get("type")
                    if bt == "thinking":
                        th = b.get("thinking")
                        if isinstance(th, str) and th.strip():
                            events.append(Event(kind="reasoning", role="assistant", text=th.strip()))
                    elif bt == "text" and isinstance(b.get("text"), str) and b["text"].strip():
                        events.append(Event(kind="assistant_msg", role="assistant", text=b["text"].strip()))
                    elif bt == "toolCall":
                        args = b.get("arguments")
                        if not isinstance(args, str):
                            args = json.dumps(args, ensure_ascii=False) if args is not None else ""
                        events.append(Event(kind="tool_call", text=b.get("name") or "?", tool_args=args))
                        if b.get("id"):
                            calls[b["id"]] = len(events) - 1
                    elif bt == "image":
                        n_img += 1
            elif isinstance(content, str) and content.strip():
                events.append(Event(kind="assistant_msg", role="assistant", text=content.strip()))
        elif role == "toolResult":
            text, n_img = _content_text(msg.get("content"))
            tid = msg.get("toolCallId")
            if tid and tid in calls:
                events[calls[tid]].tool_output = text
            else:
                events.append(Event(kind="tool_output", text=text))
                warnings.append("存在无法配对 toolCallId 的工具输出（已顺序追加）")
        elif role == "bashExecution":
            events.append(Event(
                kind="tool_call",
                text="bash",
                tool_args=msg.get("command") or "",
                tool_output=msg.get("output") or "",
            ))
        return n_img
